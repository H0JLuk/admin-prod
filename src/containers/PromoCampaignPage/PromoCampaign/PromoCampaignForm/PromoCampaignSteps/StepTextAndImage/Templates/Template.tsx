import React, { useState } from 'react';
import { BannerCreateDto, BannerCreateTextDto } from '@types';
import { Col, FormItemProps, Row } from 'antd';
import UploadPicture from '@components/UploadPicture/UploadPicture';
import { getLabel } from '@components/LabelWithTooltip/LabelWithTooltip';
import TextBlock from '../TextBlock';
import { INFO_ROWS, INFO_ROWS_KEYS, TemplateRowsValues } from './templateConstants';
import { BUTTON_TEXT } from '@constants/common';

import styles from './Template.module.css';

export type TemplateProps = {
    banners: BannerCreateDto;
    texts: BannerCreateTextDto;
    type: INFO_ROWS_KEYS;
    onRemoveImg: (name: FormItemProps['name']) => void;
    isCopy?: boolean;
};

const Template: React.FC<TemplateProps> = ({ banners, texts, type, onRemoveImg, isCopy }) => {
    const [infoRows] = useState<Record<string, TemplateRowsValues>[]>(INFO_ROWS[type] || []);

    return infoRows.map((row, index) => (
        <Row
            key={index}
            className={styles.row}
            gutter={[16, 32]}
        >
            {Object.keys(row).map((key) => (
                <Col
                    key={key}
                    className={styles.col}
                    span={12}
                >
                    {(row[key].type === 'banner' || row[key].type === 'logo') && (
                        <UploadPicture
                            description={row[key].description}
                            setting={row[key].setting}
                            accept={row[key].access_type}
                            name={['banners', key]}
                            initialValue={banners[key] as string}
                            rules={row[key].rules}
                            onRemoveImg={onRemoveImg}
                            type={row[key].type}
                            label={getLabel(row[key].title, row[key].tooltipImg, true)}
                            uploadButtonText={BUTTON_TEXT.ADD}
                            maxFileSize={row[key].maxSize}
                            validateFileSize={!isCopy}
                            footer
                        />
                    )}
                    {row[key].type === 'text' && (
                        <TextBlock
                            title={row[key].title}
                            placeholder={row[key].placeholder}
                            maxLength={row[key]?.maxLength}
                            rules={row[key].rules}
                            name={['texts', key]}
                            initialValue={texts[key]}
                        />
                    )}
                </Col>
            ))}
        </Row>
    )) as Exclude<React.ReactNode, null | undefined | boolean | React.ReactFragment>;
};

export default Template;
