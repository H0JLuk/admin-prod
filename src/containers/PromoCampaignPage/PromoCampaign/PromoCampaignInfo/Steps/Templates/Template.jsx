import React, { useState } from 'react';
import { Col, Row } from 'antd';
import { TextBlock, ImageBlock } from './TemplateBlocks';
import { INFO_ROWS } from './templateConstants';

import styles from './Template.module.css';

const getBanners = (banners) => banners.reduce(
    (result, { type, url }) => ({ ...result, [type]: url }),
    {}
);

const getTexts = (texts) => texts.reduce((result, { type, value }) => {
    if (Object.prototype.hasOwnProperty.call(result, type)) {
        return result;
    }
    return { ...result, [type]: value };
}, {});

const Template = ({ banners, texts, type }) => {
    const [filteredBanners] = useState(() => getBanners(banners));
    const [filteredTexts] = useState(() => getTexts(texts));
    const [infoRows] = useState(INFO_ROWS[type] || []);

    return infoRows.map((row, index) => (
        <Row
            key={ index }
            className={ styles.row }
            gutter={ [16] }
        >
            { Object.keys(row).map((key) => (
                <Col
                    key={ key }
                    className={ styles.col }
                    span={ 12 }
                >
                    { ['banner', 'logo'].includes(row[key].type) && (
                        <ImageBlock
                            label={ row[key].label }
                            type={ row[key].type }
                            src={ filteredBanners[key] }
                        />
                    ) }
                    { row[key].type === 'text' && (
                        <TextBlock
                            label={ row[key].label }
                            text={ filteredTexts[key] }
                        />
                    ) }
                </Col>
            )) }
        </Row>
    ));
};

export default Template;
