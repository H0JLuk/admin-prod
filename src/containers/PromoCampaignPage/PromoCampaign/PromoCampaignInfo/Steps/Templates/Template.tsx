import React, { useState } from 'react';
import { Col, Row } from 'antd';
import { TextBlock, ImageBlock } from './TemplateBlocks';
import { INFO_ROWS, IRow } from './templateConstants';
import { PromoCampaignDto } from '@types';

import styles from './Template.module.css';

type ITemplateProps = Pick<PromoCampaignDto, 'banners' | 'type' | 'texts'>;

const getBanners = (banners: ITemplateProps['banners']) => banners.reduce<Record<string, string>>(
    (result, { type, url }) => ({ ...result, [type]: url }),
    {}
);

const getTexts = (texts: ITemplateProps['texts']) => texts.reduce<Record<string, string>>((result, { type, value }) => {
    if (Object.prototype.hasOwnProperty.call(result, type)) {
        return result;
    }
    return { ...result, [type]: value };
}, {});

const Template: React.FC<ITemplateProps>= ({ banners, texts, type }) => {
    const [filteredBanners] = useState(() => getBanners(banners));
    const [filteredTexts] = useState(() => getTexts(texts));
    const [infoRows] = useState<IRow[]>(INFO_ROWS[type] || []);

    return infoRows.map((row, index) => (
        <Row
            key={index}
            className={styles.row}
            gutter={16}
        >
            {Object.keys(row).map((key) => (
                <Col
                    key={key}
                    className={styles.col}
                    span={12}
                >
                    {['banner', 'logo'].includes(row[key].type) && (
                        <ImageBlock
                            label={row[key].label}
                            type={row[key].type}
                            src={filteredBanners[key]}
                        />
                    )}
                    {row[key]?.type === 'text' && (
                        <TextBlock
                            label={row[key].label}
                            text={filteredTexts[key]}
                        />
                    )}
                </Col>
            ))}
        </Row>
    )) as Exclude<React.ReactNode, null | undefined | boolean | React.ReactFragment>;
};

export default Template;
