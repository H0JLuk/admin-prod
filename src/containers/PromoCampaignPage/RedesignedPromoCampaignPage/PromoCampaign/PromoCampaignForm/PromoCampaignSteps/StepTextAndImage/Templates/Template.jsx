import React, { useState, useEffect } from 'react';
import { Col, Row } from 'antd';
import ImgBlock from '../ImgBlock/ImgBlock';
import TextBlock from '../TextBlock/TextBlock';
import { INFO_ROWS } from './templateConstants';

import styles from './Template.module.css';

const Template = ({ banners, texts, type, onRemoveImg }) => {
    const [infoRows, setInfoRows] = useState([]);

    useEffect(() => {
        setInfoRows(INFO_ROWS[type] || []);
    }, [type]);

    return infoRows.map((row, index) => (
        <Row
            key={ index }
            className={ styles.row }
            gutter={ [16, 32] }
        >
            { Object.keys(row).map((key) => (
                <Col
                    key={ key }
                    className={ styles.col }
                    span={ 12 }
                >
                    { ['logo', 'banner'].includes(row[key].type) && (
                        <ImgBlock
                            title={ row[key].title }
                            description={ row[key].description }
                            setting={ row[key].setting }
                            tooltipImg={ row[key].tooltipImg }
                            access_type={ row[key].access_type }
                            name={ `promoCampaignBanners.${key}` }
                            initialValue={ banners[key] }
                            rules={ row[key].rules }
                            onRemoveImg={ onRemoveImg }
                            type={ row[key].type }
                        />
                    ) }
                    { row[key].type === 'text' && (
                        <TextBlock
                            title={ row[key].title }
                            placeholder={ row[key].placeholder }
                            maxLength={ row[key]?.maxLength }
                            rules={ row[key].rules }
                            name={ `promoCampaignTexts.${key}` }
                            initialValue={ texts[key] }
                        />
                    ) }
                </Col>
            )) }
        </Row>
    ));
};

export default Template;