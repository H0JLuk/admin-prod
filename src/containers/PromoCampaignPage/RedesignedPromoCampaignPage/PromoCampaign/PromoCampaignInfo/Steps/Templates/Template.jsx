import React, { useState, useEffect } from 'react';
import { Col, Row } from 'antd';
import { TextBlock, ImageBlock } from './TemplateBlocks';
import { INFO_ROWS } from './templateConstants';

import styles from './Template.module.css';

const Template = ({ banners, texts, type }) => {
    const [filteredBanners, setFilteredBanners] = useState({});
    const [filteredTexts, setFilteredTexts] = useState({});
    const [infoRows, setInfoRows] = useState([]);

    useEffect(() => {
        const resultBanners = banners.reduce((result, { type, url }) => {
            if (Object.prototype.hasOwnProperty.call(result, type)) {
                return result;
            }
            return { ...result, [type]: url };
        }, {});
        const resultTexts = texts.reduce((result, { type, value }) => {
            if (Object.prototype.hasOwnProperty.call(result, type)) {
                return result;
            }
            return { ...result, [type]: value };
        }, {});

        setInfoRows(INFO_ROWS[type] || []);
        setFilteredBanners(resultBanners);
        setFilteredTexts(resultTexts);
    }, [banners, texts, type]);

    return infoRows.map((row, index) => (
        <Row
            key={ index }
            className={ styles.row }
            gutter={ [16, 32] }
        >
            { Object.keys(row).map((key) => (
                <Col
                    key={ key }
                    className={ styles.col }
                    span={ 12 }
                >
                    { ['image', 'icon'].includes(row[key].type) && (
                        <ImageBlock
                            label={ row[key].label }
                            type={ row[key].type }
                            src={ filteredBanners[key] }
                        />
                    ) }
                    { row[key].type === 'text' && (
                        <TextBlock
                            label={ row[key].label }
                            text={ filteredTexts[key] }
                        />
                    ) }
                </Col>
            )) }
        </Row>
    ));
};

export default Template;