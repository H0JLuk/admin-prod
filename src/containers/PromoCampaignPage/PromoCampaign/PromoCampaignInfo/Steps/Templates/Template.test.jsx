import React from 'react';
import Template from './Template';
import { shallow } from 'enzyme';
import {
    getTextsReturnValue,
    promoCampaignBannerArray,
    promoCampaignTextsArray,
} from '../../../../../../../__tests__/constants';

describe('<Template /> test', () => {
    const props = {
        banners: promoCampaignBannerArray,
        texts: promoCampaignTextsArray,
        type: 'NORMAL' // 'NORMAL' || 'PRESENT'
    };

    it('Template snapshot', () => {
        const TemplateItem = shallow(<Template { ...props } />);
        expect(TemplateItem.first().html()).toMatchSnapshot();
    });

    it('should render ImageBlock and TextBlock', () => {
        const TemplateItem = shallow(<Template { ...props } />);

        expect(TemplateItem.find('ImageBlock').first()).toHaveLength(1);
        expect(TemplateItem.find('TextBlock').first()).toHaveLength(1);
    });

    it('should not render Template', () => {
        const newProps = {
            ...props,
            type: 'errorType'
        };

        const TemplateItem = shallow(<Template { ...newProps } />);
        expect(TemplateItem).toHaveLength(0);
    });

    it('should return only result in getTexts func', () => {
        const newProps = {
            ...props,
            texts: getTextsReturnValue
        };

        const TemplateItem = shallow(<Template { ...newProps } />);
        expect(TemplateItem.find('TextBlock').first()).toHaveLength(1);
    });
});
