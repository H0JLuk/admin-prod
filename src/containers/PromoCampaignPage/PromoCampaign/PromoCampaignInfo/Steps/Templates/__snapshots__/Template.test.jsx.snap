// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Template /> test Template snapshot 1`] = `"<div class=\\"ant-row row\\" style=\\"margin-left:-8px;margin-right:-8px\\"><div style=\\"padding-left:8px;padding-right:8px\\" class=\\"ant-col ant-col-12 col\\"><div class=\\"title\\">Баннер на главной</div><div class=\\"block\\"><div class=\\"\\"><img class=\\"banner\\" src=\\"http://distributor-fs:8081/distributor-fs/file?path=greenday-presents/promocampaign/sberhealth/2/card.jpg\\" alt=\\"\\"/></div><div class=\\"footer\\"><div class=\\"fileName\\">card.jpg</div><div class=\\"fileSize\\"></div></div></div></div><div style=\\"padding-left:8px;padding-right:8px\\" class=\\"ant-col ant-col-12 col\\"><div class=\\"title\\">Логотип на экране с номером (белый)</div><div class=\\"block\\"><div class=\\"logo\\"><img class=\\"imageLogo\\" alt=\\"\\"/></div><div class=\\"footer\\"><div class=\\"fileName\\"></div><div class=\\"fileSize\\"></div></div></div></div></div>"`;
